InnerEye/ML/SSL/lightning_modules/byol/byol_module.py

```diff
@@ -0,0 +1,164 @@
+                 encoder_name: str,
+                 warmup_epochs: int,
+                 use_7x7_first_conv_in_resnet: bool = True,
+                 weight_decay: float = 1e-6,
+                 **kwargs: Any) -> None:
+        """
+        Args:
+            :param num_samples: Number of samples present in training dataset / dataloader.
+            :param learning_rate: Optimizer learning rate.
+            :param batch_size: Sample batch size used in gradient updates.
+            :param encoder_name: Type of CNN encoder used to extract image embeddings. The options are:
+                          {'resnet18', 'resnet50', 'resnet101', 'densenet121'}.
+            :param warmup_epochs: Number of epochs for scheduler warm up (linear increase from 0 to base_lr).
+            :param use_7x7_first_conv_in_resnet: If True, use a 7x7 kernel (default) in the first layer of resnet.
+            If False, replace first layer by a 3x3 kernel. This is required for small CIFAR 32x32 images to not
+            shrink them.
+            :param weight_decay: L2-norm weight decay.
+        """
+        super().__init__()
+        self.save_hyperparameters()
+
+        self.min_learning_rate = 1e-4
+        self.online_network = SiameseArm(encoder_name, use_7x7_first_conv_in_resnet)
+        self.target_network = deepcopy(self.online_network)
+        self.weight_callback = ByolMovingAverageWeightUpdate()
+
+    def on_train_batch_end(self, *args: Any, **kwargs: Any) -> None:
+        # Add callback for user automatically since it's key to BYOL weight update
+        self.weight_callback.on_before_zero_grad(self.trainer, self)
+
+    def forward(self, x: T) -> T:  # type: ignore
+        return self.target_network.encoder(x)
+
+    @staticmethod
+    def cosine_loss(a: T, b: T) -> T:
+        a = F.normalize(a, dim=-1)
+        b = F.normalize(b, dim=-1)
+        neg_cos_sim = -(a * b).sum(dim=-1).mean()
+        return neg_cos_sim
+
+    def shared_step(self, batch: BatchType, batch_idx: int) -> T:
+        """
+        Returns the BYOL loss for a given batch of images, used in validation
+        and training step.
+        :param batch: assumed to be a batch a Tuple(List[tensor, tensor, tensor], tensor) to match lightning-bolts
+        SimCLRTrainDataTransform API; the first tuple element contains a list of three tensor where the two first
+        elements contain two are two strong augmented versions  of the original images in the batch and the last
+        is a milder augmentation (ignored here).
+        :param batch_idx: index of the batch
+        :return: BYOL loss
+        """
+        batch = batch[SSLDataModuleType.ENCODER] if isinstance(batch, dict) else batch
+        (img_1, img_2), _ = batch
+
+        # Image 1 to image 2 loss
+        _, _, h_img1 = self.online_network(img_1)
+        _, _, h_img2 = self.online_network(img_2)
+        with torch.no_grad():
+            _, z_img1, _ = self.target_network(img_1)
```

Comment:
we could actually reduce compute time by adding a new method in `SiameseArm` that only propagates until encoding and doesn't do any projection. It will be a minor improvement but it requires only a small change.

Reply:
updated good point